urlcaef331f4e_dreamable_cra_tgz-pJ8-coho_lp_project_LPProjectJ8/src/coho/interp/BasicFunctions.java

```diff
@@ -0,0 +1,216 @@
+			public Object nextElement() { return(functions[i++]); }});
+	}
+
+	private static Function[] functions = new Function[] {
+		new Function() {
+			public String name() { return("print"); }
+			public Value eval(RCvalue args) throws EvalException {
+				do_print(args, "");
+				return(VoidValue.instance());
+			}
+		},
+		new Function() {
+			public String name() { return("println"); }
+			public Value eval(RCvalue args) throws EvalException {
+				do_print(args, "\n");
+				return(VoidValue.instance());
+			}
+		},
+		new Function() {
+			public String name() { return("exit"); }
+			public Value eval(RCvalue args) throws EvalException {
+				if(STAT.stat)STAT.outStat();//XXX for stat
+				if(args.size() == 0) System.exit(0);
+				else if(args.size() == 1)
+					System.exit(BasicArgs.int_arg(args, 0, name()));
+				else throw new EvalException(name() +
+						": must be invoked with no arguments, or with one" +
+				" argument that is a double.");
+				return(VoidValue.instance());
+			}
+		},
+		new Function(){
+			public String name(){return("step");}
+			public Value eval(RCvalue args)throws EvalException{
+				if(args.size()!=1)
+					throw new EvalException(name()+" step must be invoked with one int arguments");
+				int step = BasicArgs.int_arg(args,0,name());
+				//TODO: do nothing when running(it possible distrub the system.out); only for stat, use the log file to generate the data again. 
+				if(STAT.stat) STAT.outStepStat(step);
+				return (VoidValue.instance());
+			}
+		},
+		new Function(){
+			public String name(){return("set");}
+			public Value eval(RCvalue args)throws EvalException{
+				if(args.size()!=2)
+					throw new EvalException(name()+"set must be invoked with two arguments");
+				int key = BasicArgs.int_arg(args, 0, name());
+				double value = BasicArgs.double_arg(args, 1, name());
+				switch(key){
+				case 1:
+					OutwardReduce.maxNewEdgeLen = value;
+					break;
+				case 2:
+					OutwardReduce.maxNewPointDist = value; break;
+				default:
+					throw new EvalException(name()+"set function does not support the key "+key);
+				}
+				return (VoidValue.instance());
+			}
+		},
+		new Function(){
+			public String name(){return("get");}
+			public Value eval(RCvalue args)throws EvalException{
+				if(args.size()!=1)
+					throw new EvalException(name()+"get must be invoked with one arguments");
+				int key = BasicArgs.int_arg(args, 0, name());
+				switch(key){
+				case 1:
+					return new DoubleValue(OutwardReduce.maxNewEdgeLen); 
+				case 2:
+					return new DoubleValue(OutwardReduce.maxNewPointDist); 
+				default:
+					throw new EvalException(name()+"set function does not support the key "+key);
+				}
+			}
+		},
+		new Function() {
+			public String name() { return("load"); }
+			public Value eval(RCvalue args) throws EvalException {
+				String fname = BasicArgs.string_arg(args, 0, name());
+				FileReader rd;
+				try { rd = new FileReader(fname); }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *rd*. Currently, there are execution paths that do not contain closure statements, for example, when an if condition at line 121 evaluates to true. Ensure *rd* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).